misc/async-bounded-queue/src/lib.rs

```diff
@@ -0,0 +1,88 @@
+use std::collections::VecDeque;
+use std::task::{Context, Poll, Waker};
+
+/// A bounded queue for use in asynchronous contexts.
+pub struct Queue<T> {
+    storage: VecDeque<T>,
+
+    queue_full_waker: Option<Waker>,
+    queue_empty_waker: Option<Waker>,
+}
+
+impl<T> Queue<T> {
+    /// Construct a queue with the given capacity.
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            storage: VecDeque::with_capacity(capacity),
```

Comment:
Just a thought: Say that I am using `Queue` with a large non-boxed `T` type. Say that I want a bound of some large number, e.g. 1_000. Say that in the usual case, I only ever have 10 elements here. In such case, pre-allocating the whole vector would be a waste of memory.

Does the above make sense? Do we even need to care about this edge-case? I am happy to delay this decision.

Reply:
I concur that this might be an issue — perhaps it is better to limit the allocation of the initial capacity to some (configurable?) value. Or perhaps more intuitive, add a second parameter for the initial capacity.

OTOH: if there are such knobs on the facility, people would probably expect the capacity to shrink back to its “desired value” after peak usage. A fixed buffer will always continue to work after initial allocation, but a variable buffer (or rather: thousands thereof) can cause more erratic runtime problems.

Reply:
I'd like to avoid multiple configuration parameters if possible. What about no allocating initial capacity and calling `shrink_to_fit` as part of `poll_pop`?

Reply:
That would keep the memory usage as low as possible.